builder/transaction_builder.go

```diff
@@ -676,62 +690,131 @@ func (b *TransactionBuilder) Build(signer iotago.AddressSigner) (*iotago.SignedT
 		Transaction: b.transaction,
-		Unlocks:     unlocks,
+		Unlocks:     unlockedSet.unlocks,
 	}
 
 	return sigTxPayload, nil
 }
 
-func addReferentialUnlock(addr iotago.Address, unlocks iotago.Unlocks, pos int) iotago.Unlocks {
+// txBuilderUnlockedSet is a helper struct to keep track of the unlocked inputs and their positions.
+type txBuilderUnlockedSet struct {
+	// unlocks holds the unlocks for the tx inputs.
+	unlocks iotago.Unlocks
+	// signerUIDs maps unique signer UIDs to the position of the unlock in the unlocks slice.
+	signerUIDs map[iotago.Identifier]int
+	// unlockedChains maps the chain address key to the position of the unlock in the unlocks slice.
+	unlockedChains map[string]int
+}
+
+// addUnlock adds the given unlock to the set.
+func (u *txBuilderUnlockedSet) addUnlock(unlock iotago.Unlock) {
+	u.unlocks = append(u.unlocks, unlock)
+}
+
+// addReferentialUnlock adds a referential unlock to the set.
+func (u *txBuilderUnlockedSet) addReferentialUnlock(addr iotago.Address, referencedInputIndex int) {
 	switch addr.(type) {
 	case *iotago.AccountAddress:
-		return append(unlocks, &iotago.AccountUnlock{Reference: uint16(pos)})
+		u.addUnlock(&iotago.AccountUnlock{Reference: uint16(referencedInputIndex)})
 	case *iotago.AnchorAddress:
-		return append(unlocks, &iotago.AnchorUnlock{Reference: uint16(pos)})
+		u.addUnlock(&iotago.AnchorUnlock{Reference: uint16(referencedInputIndex)})
 	case *iotago.NFTAddress:
-		return append(unlocks, &iotago.NFTUnlock{Reference: uint16(pos)})
+		u.addUnlock(&iotago.NFTUnlock{Reference: uint16(referencedInputIndex)})
 	default:
-		return append(unlocks, &iotago.ReferenceUnlock{Reference: uint16(pos)})
+		u.addUnlock(&iotago.ReferenceUnlock{Reference: uint16(referencedInputIndex)})
 	}
 }
 
-func addChainAsUnlocked(input iotago.Output, posUnlocked int, prevUnlocked map[string]int) {
+// addSignerUID marks the given signer UID as unlocked at "unlockedAtIndex".
+func (u *txBuilderUnlockedSet) addSignerUID(signerUID iotago.Identifier, unlockedAtIndex int) {
+	u.signerUIDs[signerUID] = unlockedAtIndex
+}
+
+// addChainAsUnlocked marks the underlying chain address as unlocked at "unlockedAtIndex".
+// This is only done if the output is a chain output and the chain ID is addressable,
+// which is only valid for AccountOutput, AnchorOutput and NFTOutput.
+func (u *txBuilderUnlockedSet) addChainAsUnlocked(input iotago.Output, unlockedAtIndex int) {
 	if chainInput, is := input.(iotago.ChainOutput); is && chainInput.ChainID().Addressable() {
-		prevUnlocked[chainInput.ChainID().ToAddress().Key()] = posUnlocked
+		u.unlockedChains[chainInput.ChainID().ToAddress().Key()] = unlockedAtIndex
 	}
 }
+
+// isChainUnlocked checks if the underlying chain address is unlocked and returns the unlock index.
+func (u *txBuilderUnlockedSet) isChainUnlocked(chainAddr iotago.ChainAddress) (int, bool) {
+	unlockedAtIndex, isUnlocked := u.unlockedChains[chainAddr.ChainID().ToAddress().Key()]
+	return unlockedAtIndex, isUnlocked
```

Comment:
return directly

Reply:
Unfortunately the go compiler doesn't understand that it needs to return two values if I return directly.
I will leave it as it is.